enclave-runtime/src/lib.rs

```diff
@@ -497,78 +546,95 @@ where
-		return None
+		return Err(Error::Stf("bad signature".to_string()))
 	}
 
 	if !state::exists(&shard) {
 		info!("Initialized new shard that was queried chain: {:?}", shard);
 		if let Err(e) = state::init_shard(&shard) {
-			error!("Error initialising shard {:?} state: Error: {:?}", shard, e);
-			return None
+			return Err(Error::Stf(format!(
+				"Error initialising shard {:?} state: Error: {:?}",
+				shard, e
+			)))
 		}
 	}
 
 	let mut state = match state::load(&shard) {
 		Ok(s) => s,
-		Err(e) => {
-			error!("Error loading shard {:?}: Error: {:?}", shard, e);
-			return None
-		},
+		Err(e) =>
+			return Err(Error::Stf(format!("Error loading shard {:?}: Error: {:?}", shard, e))),
 	};
 
 	debug!("calling into STF to get state");
-	Stf::get_state(&mut state, trusted_getter_signed.into())
+	Ok(Stf::get_state(&mut state, trusted_getter_signed.into()))
 }
 
-fn execute_top_pool_calls<O>(
+fn execute_top_pool_calls<O, T>(
 	ocall_api: &O,
+	top_pool_getter: &T,
 	latest_onchain_header: Header,
 ) -> Result<(Vec<OpaqueCall>, Vec<SignedSidechainBlock>)>
 where
-	O: EnclaveOnChainOCallApi + EnclaveRpcOCallApi,
+	O: EnclaveOnChainOCallApi,
+	T: GetTopPool,
 {
 	debug!("Executing pending pool operations");
 
-	// load top pool
-	let pool_mutex: &SgxMutex<BPool> = match rpc::worker_api_direct::load_top_pool() {
-		Some(mutex) => mutex,
-		None => {
-			error!("Could not get mutex to pool");
-			return Error::Sgx(sgx_status_t::SGX_ERROR_UNEXPECTED).into()
-		},
-	};
-
-	let pool_guard: SgxMutexGuard<BPool> = pool_mutex.lock().unwrap();
-	let pool: Arc<&BPool> = Arc::new(pool_guard.deref());
-	let author: Arc<Author<&BPool>> = Arc::new(Author::new(pool.clone()));
+	// // load top pool
+	// let pool_mutex = match top_pool_getter.get() {
+	// 	Some(mutex) => mutex,
+	// 	None => {
+	// 		error!("Could not get mutex to pool");
+	// 		return Error::Sgx(sgx_status_t::SGX_ERROR_UNEXPECTED).into()
+	// 	},
+	// };
+	//
+	// let pool: Arc<&BasicPool<T::SideChainApi, T::Block, T::RpcResponder>> =
+	// 	Arc::new(pool_mutex.lock().unwrap().deref());
+	// let author: Arc<Author<&BasicPool<T::SideChainApi, T::Block, T::RpcResponder>>> =
+	// 	Arc::new(Author::new(pool.clone()));
 
 	// get all shards
 	let shards = state::list_shards()?;
 
 	// Handle trusted getters
-	execute_trusted_getters(ocall_api, &author, &shards)?;
+	execute_trusted_getters(top_pool_getter, &shards)?;
 
 	// Handle trusted calls
 	let calls_and_blocks =
-		execute_trusted_calls(ocall_api, latest_onchain_header, pool, author, shards)?;
+		execute_trusted_calls(ocall_api, latest_onchain_header, top_pool_getter, shards)?;
 
 	Ok(calls_and_blocks)
 }
 
-fn execute_trusted_calls<O>(
+fn execute_trusted_calls<O, T>(
 	on_chain_ocall: &O,
 	latest_onchain_header: Header,
-	pool: Arc<&BPool>,
-	author: Arc<Author<&BPool>>,
+	top_pool_getter: &T,
 	shards: Vec<H256>,
 ) -> Result<(Vec<OpaqueCall>, Vec<SignedSidechainBlock>)>
 where
 	O: EnclaveOnChainOCallApi,
+	T: GetTopPool,
```

Comment:
pass in the top pool getter, so we don't have to access a global variable inside this function. makes it better testable.

Reply:
Here I am not certain if I agree. Before, we passed in a pool type directly without any kind of notion of globality; hence, we could pass in any kind of prepared pool for testing. 

Now with the `top_pool_getter`, we gave the function something to access state outside the function's scope. So, I actually think it is a step backwards for testing. Am I missing something?

Reply:
`GetTopPool` is just a proxy - this function does not care where the pool comes from. This can be a global context or a local one (as is the case in the tests that call this function). It's an additional layer of indirection and generalization (because we don't want a regular `BPool` in our test cases)

Reply:
Hmm, I understand that we not necessarily want a regular pool (although in the pool case, we can imo use the regular one). However, I prefer the function be abstracted over the pool traits directly instead of the indirect getter.

I think that layers of indirections add unnecessary complexity when they are not strictly needed.

Reply:
The 'regular' pool uses an RPC response sender that requires a real WS connection in the background, which is not what we want in test cases, so we cannot use the regular `BPool` for testing.

Regarding the getter vs passing the author and pool directly: We can get rid of the getter indirection, but then I'd suggest to just pass in the pool and construct the author inside the function (reduces the surface area of the function)

Reply:
Ah ok sorry. I wasn't aware of the pool's complexity. And if we have not completely abstracted the pool behind a trait yet, I am totally fine with the getter approach.

> Regarding the getter vs passing the author and pool directly: We can get rid of the getter indirection, but then I'd suggest to just pass in the pool and construct the author inside the function (reduces the surface area of the function) 

Yes I agree, I tackled that different in my branch. I simply added a `fn pool()` getter in the author to access the pool from the author.

Reply:
I was also tempted to make the pool accessible through the author, but it undermines the abstraction layer of the author. So I think it's cleaner to construct an author from a pool where needed and then use it only as author (especially if we abstract the author in the future, it will be painful to have such a 'breach' into internals of the author, which is exactly what I had with the `TrustedOperationPool` and got rid of now).

I updated this part of the code now, I think it's quite a bit cleaner now, thanks to our discussion 👍